File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l1_cur = l1
        l2_cur = l2
        carry = 0
        
        new_node = ListNode()
        start = new_node
        prev_node = None
        while l1_cur is not None or l2_cur is not None:
            l1_cur_val = l2_cur_val = 0
            if l1_cur:
                l1_cur_val = l1_cur.val
            if l2_cur:
                l2_cur_val = l2_cur.val
            new_val = l1_cur_val + l2_cur_val + carry
            if new_val >= 10:
                carry = 1
                new_val -= 10
            else:
                carry = 0
            new_node.val = new_val
            if prev_node:
                prev_node.next = new_node
            prev_node = new_node
            new_node = ListNode()
            if l1_cur:
                l1_cur = l1_cur.next
            if l2_cur:
                l2_cur = l2_cur.next
        if carry == 1:
            new_node = ListNode(1)
            prev_node.next = new_node
        return start
        
        
        

File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1->val == 0 && !l1->next) return l2;
        if (l2->val == 0 && !l2->next) return l1;
        
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        ListNode* cur = new ListNode();
        ListNode* head = cur;
        int carry = 0;
        
        while (cur1 || cur2){
            cur->next = new ListNode();
            cur = cur->next;
            
            int newDigit = carry;
            if (cur1){
                newDigit += cur1->val;
                cur1 = cur1->next;
            }
            
            if (cur2){
                newDigit += cur2->val;
                cur2 = cur2->next;
            }
            cur->val = newDigit % 10;
            carry = newDigit / 10;
            
        }
        if(carry == 1) cur->next = new ListNode(1);
        ListNode* ans = head->next;
        delete head;
        return ans;
        
    }
};